### PR TITLE
Ignore unattached variables configs

### PR DESCRIPTION
**Pull request by webrouse**
Part of: https://keboola.atlassian.net/browse/KAC-49

Changes:
- All remote unattached variables config (no config uses they) are ignored.

### CODE REVIEW

File: internal/pkg/mapper/ignore/ignore.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import (
+	"github.com/keboola/keboola-as-code/internal/pkg/log"
+	"github.com/keboola/keboola-as-code/internal/pkg/state"
+)
+
+type ignoreMapper struct {
+	state  *state.State
+	logger log.Logger
+}
+
```
Comment (webrouse) [inline internal/pkg/mapper/ignore/ignore.go:12]: novy `mappers/ignore` na ignorovanie nejakych objektov, ktore este nemame pokryte, ... aktualne iba variables.

File: internal/pkg/mapper/ignore/remote.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+// isIgnoredConfig ignores all variables configs which are not attached to a config.
+func (m *ignoreMapper) isIgnoredConfig(config *model.Config) bool {
+	// Variables config
+	if config.ComponentId != model.VariablesComponentId {
+		return false
+	}
+
+	// Without target config
+	if !config.Relations.Has(model.VariablesForRelType) && !config.Relations.Has(model.SharedCodeVariablesForRelType) {
```
Comment (webrouse) [inline internal/pkg/mapper/ignore/remote.go:65]: Ignorovana je variables configuration, ktora nie je napojena na ziadny config.

File: internal/pkg/mapper/ignore/remote_test.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	}
+	assert.NoError(t, state.Set(unattachedVars))
+
+	// Invoke
+	changes := model.NewRemoteChanges()
+	changes.AddLoaded(targetConfig)
+	changes.AddLoaded(targetVars)
+	changes.AddLoaded(unattachedVars)
+	assert.NoError(t, state.Mapper().AfterRemoteOperation(changes))
+	assert.Equal(t, "DEBUG  Ignored unattached variables config \"branch:1/component:keboola.variables/config:3\"\n", logger.AllMessages())
```
Comment (webrouse) [inline internal/pkg/mapper/ignore/remote_test.go:54]: Unit test.

File: internal/pkg/model/changes.go [hunk 4]
```diff
@@ -113,21 +113,11 @@ func (c *LocalChanges) AddRenamed(actions ...RenameAction) {
...
-		c.updated[i] = callback(v)
-	}
-	for i, v := range c.saved {
-		c.saved[i] = callback(v)
-	}
+	c.loaded = replaceObjects(c.loaded, callback)
+	c.persisted = replaceObjects(c.persisted, callback)
+	c.created = replaceObjects(c.created, callback)
+	c.updated = replaceObjects(c.updated, callback)
+	c.saved = replaceObjects(c.saved, callback)
```
Comment (webrouse) [inline internal/pkg/model/changes.go:120]: Pouzita spolocna funkcia `replaceObjects`, ktora ignoruje `nil` hodnotu, ak chceme zo zoznamu nieco vyhodit.

File: internal/pkg/mapper/ignore/remote.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+				object.SetRemoteState(nil)
+			} else {
+				// Remove state
+				m.state.Remove(object.Key())
+			}
+		}
+	}
+
+	// Fix list of the changed objects
+	changes.Replace(func(v model.ObjectState) model.ObjectState {
```
Comment (webrouse) [inline internal/pkg/mapper/ignore/remote.go:26]: Odstranene objekty je potrebne aj vyhodit zo zoznamu zmien.

File: internal/pkg/project/mappers.go [hunk 6]
```diff
@@ -35,6 +36,8 @@ func MappersFor(s *state.State, d dependencies) (mapper.Mappers, error) {
 		sharedcode.NewLinksMapper(s),
 		// Relations between objects
 		relations.NewMapper(s),
+		// Skip variables configurations that are not used in any configuration
+		ignore.NewMapper(s),
```
Comment (webrouse) [inline internal/pkg/project/mappers.go:40]: Mapper sa spusta v `template/project` po `relations` mapper, takze `relations` su uz nastavene na oboch stranach.

File: internal/pkg/state/remote/manager.go [hunk 7]
```diff
@@ -343,6 +343,7 @@ func (u *UnitOfWork) createRequest(objectState model.ObjectState, object model.O
 					if r, err := u.updateRequest(objectState, object, recipe, nil); err != nil {
 						response.SetErr(err)
 					} else {
+						response.SetErr(nil)
```
Comment (webrouse) [inline internal/pkg/state/remote/manager.go:346]: Cely kod:
https://github.com/keboola/keboola-as-code/blob/32dc410f42d3d1a08678abb9ca33677460d42069/internal/pkg/state/remote/manager.go#L339-L352

Riesi sa tu edge-case, ze ak config/row uz existuje (napr. vznikol behom requestu), alebo nejaka edge case v API, tak sa pouzije update namiesto create.

Mal som chybu v kode a spustil sa tento kod, .... `update` request sa spusti, ale `err` sa nevymazal, takze opracia spadla.

File: test/pull/variables-rm-parent/expected-stdout [hunk 8]
```diff
@@ -1,10 +1,8 @@
 Plan for "pull" operation:
   × C my-branch/transformation/keboola.snowflake-transformation/snowflake-transformation
-  * C my-branch/transformation/keboola.snowflake-transformation/snowflake-transformation/variables | changed: relations
-  * R my-branch/transformation/keboola.snowflake-transformation/snowflake-transformation/variables/values/default | changed: relations
+  × C my-branch/transformation/keboola.snowflake-transformation/snowflake-transformation/variables
+  × R my-branch/transformation/keboola.snowflake-transformation/snowflake-transformation/variables/values/default
```
Comment (webrouse) [inline test/pull/variables-rm-parent/expected-stdout:4]: Zmeny v testoch, `variables` ktore nie su napojene na objekt, ako keby ani neexistovali.

**Review by webrouse — COMMENTED** [review]
Posledna vec, ktoru je potrebne upravit k tomuto tasku.

Ak zmazes transformaciu v UI a pouzivala variables, tak tie sa nezmazu.
Casom z toho moze byt dost vela suborov, ktore vobec nepotrebujes.
Preto som pridal, aby sa vsetky variables, ktore nie su napojene na nejaky config ignorovali.

**Review by JakubMatejka — APPROVED** [review]
